// @fluxerjs/types/dist/routes.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/subpath-routes.ts
var subpath_routes_exports = {};
__export(subpath_routes_exports, {
  Routes: () => Routes
});
module.exports = __toCommonJS(subpath_routes_exports);

// src/rest/routes.ts
var Routes = {
  // Channels
  channel: (id) => `/channels/${id}`,
  channelMessages: (id) => `/channels/${id}/messages`,
  channelMessage: (channelId, messageId) => `/channels/${channelId}/messages/${messageId}`,
  channelMessageReactions: (channelId, messageId) => `/channels/${channelId}/messages/${messageId}/reactions`,
  channelMessageReaction: (channelId, messageId, emoji) => `/channels/${channelId}/messages/${messageId}/reactions/${encodeURIComponent(emoji)}`,
  channelPins: (id) => `/channels/${id}/messages/pins`,
  /** Use channelPinMessage for PUT/DELETE pin operations. */
  channelPin: (channelId, messageId) => `/channels/${channelId}/messages/pins/${messageId}`,
  /** Pin/unpin: PUT or DELETE /channels/{id}/pins/{messageId}. Use for pin and unpin operations. */
  channelPinMessage: (channelId, messageId) => `/channels/${channelId}/pins/${messageId}`,
  channelBulkDelete: (id) => `/channels/${id}/messages/bulk-delete`,
  channelWebhooks: (id) => `/channels/${id}/webhooks`,
  channelTyping: (id) => `/channels/${id}/typing`,
  channelRtcRegions: (id) => `/channels/${id}/rtc-regions`,
  channelSlowmode: (id) => `/channels/${id}/slowmode`,
  channelInvites: (id) => `/channels/${id}/invites`,
  channelPermission: (channelId, overwriteId) => `/channels/${channelId}/permissions/${overwriteId}`,
  channelRecipient: (channelId, userId) => `/channels/${channelId}/recipients/${userId}`,
  channelMessageAttachment: (channelId, messageId, attachmentId) => `/channels/${channelId}/messages/${messageId}/attachments/${attachmentId}`,
  // Guilds
  guilds: () => "/guilds",
  guild: (id) => `/guilds/${id}`,
  guildDelete: (guildId) => `/guilds/${guildId}/delete`,
  guildVanityUrl: (guildId) => `/guilds/${guildId}/vanity-url`,
  guildTextChannelFlexibleNames: (guildId) => `/guilds/${guildId}/text-channel-flexible-names`,
  guildDetachedBanner: (guildId) => `/guilds/${guildId}/detached-banner`,
  guildDisallowUnclaimedAccounts: (guildId) => `/guilds/${guildId}/disallow-unclaimed-accounts`,
  guildTransferOwnership: (guildId) => `/guilds/${guildId}/transfer-ownership`,
  guildRolesHoistPositions: (guildId) => `/guilds/${guildId}/roles/hoist-positions`,
  guildEmojisBulk: (guildId) => `/guilds/${guildId}/emojis/bulk`,
  guildStickersBulk: (guildId) => `/guilds/${guildId}/stickers/bulk`,
  guildChannels: (id) => `/guilds/${id}/channels`,
  guildMembers: (id) => `/guilds/${id}/members`,
  guildMembersSearch: (id) => `/guilds/${id}/members-search`,
  guildMember: (guildId, userId) => `/guilds/${guildId}/members/${userId}`,
  guildMemberRole: (guildId, userId, roleId) => `/guilds/${guildId}/members/${userId}/roles/${roleId}`,
  guildRoles: (id) => `/guilds/${id}/roles`,
  guildRole: (guildId, roleId) => `/guilds/${guildId}/roles/${roleId}`,
  guildBans: (id) => `/guilds/${id}/bans`,
  guildBan: (guildId, userId) => `/guilds/${guildId}/bans/${userId}`,
  guildInvites: (id) => `/guilds/${id}/invites`,
  invite: (code) => `/invites/${encodeURIComponent(code)}`,
  guildAuditLogs: (id) => `/guilds/${id}/audit-logs`,
  guildEmojis: (id) => `/guilds/${id}/emojis`,
  guildEmoji: (guildId, emojiId) => `/guilds/${guildId}/emojis/${emojiId}`,
  guildStickers: (id) => `/guilds/${id}/stickers`,
  guildSticker: (guildId, stickerId) => `/guilds/${guildId}/stickers/${stickerId}`,
  guildWebhooks: (id) => `/guilds/${id}/webhooks`,
  webhook: (id) => `/webhooks/${id}`,
  webhookExecute: (id, token) => `/webhooks/${id}/${token}`,
  webhookMessage: (id, token, messageId) => `/webhooks/${id}/${token}/messages/${messageId}`,
  // Users
  user: (id) => `/users/${id}`,
  currentUser: () => `/users/@me`,
  currentUserGuilds: () => `/users/@me/guilds`,
  leaveGuild: (guildId) => `/users/@me/guilds/${guildId}`,
  userMeChannels: () => `/users/@me/channels`,
  /** GET /users/{id}/profile. Pass guildId for server-specific profile. */
  userProfile: (id, guildId) => guildId ? `/users/${id}/profile?guild_id=${guildId}` : `/users/${id}/profile`,
  // Instance (unauthenticated)
  /** Canonical instance discovery document (may mirror GET /instance). */
  instanceDiscovery: () => "/.well-known/fluxer",
  instance: () => "/instance",
  // Gateway
  gatewayBot: () => `/gateway/bot`,
  // Streams (voice channel screen share preview)
  streamPreview: (streamKey) => `/streams/${encodeURIComponent(streamKey)}/preview`,
  // Application commands (slash commands)
  applicationCommands: (applicationId) => `/applications/${applicationId}/commands`,
  applicationCommand: (applicationId, commandId) => `/applications/${applicationId}/commands/${commandId}`,
  interactionCallback: (interactionId, interactionToken) => `/interactions/${interactionId}/${interactionToken}/callback`,
  // OAuth2 / Bot
  oauth2ApplicationBot: (id) => `/oauth2/applications/${id}/bot`,
  oauth2ApplicationBotResetToken: (id) => `/oauth2/applications/${id}/bot/reset-token`
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Routes
});
