// @fluxerjs/core/dist/message.mjs
// src/structures/Base.ts
var Base = class {
};

// src/structures/Message.ts
import { Collection as Collection2 } from "@fluxerjs/collection";
import { MessageType, Routes } from "@fluxerjs/types";
import { EmbedBuilder as EmbedBuilder2 } from "@fluxerjs/builders";

// src/util/messageUtils.ts
import { MessageFlags } from "@fluxerjs/types";
import { EmbedBuilder } from "@fluxerjs/builders";

// src/errors/ErrorCodes.ts
import { ErrorCodes } from "@fluxerjs/util";

// src/errors/FluxerError.ts
import { FluxerError } from "@fluxerjs/util";

// src/util/messageUtils.ts
var FILE_FETCH_TIMEOUT_MS = 3e4;
async function resolveMessageFiles(files) {
  const result = [];
  for (let i = 0; i < files.length; i++) {
    const f = files[i];
    if (!f) continue;
    const filename = f.filename ?? f.name;
    if ("url" in f && f.url) {
      if (!URL.canParse(f.url)) {
        throw new FluxerError(`Invalid file URL at index ${i}: ${f.url}`, {
          code: ErrorCodes.InvalidFileUrl
        });
      }
      let res;
      try {
        res = await fetch(f.url, {
          signal: AbortSignal.timeout(FILE_FETCH_TIMEOUT_MS)
        });
      } catch (err) {
        throw new FluxerError(`Failed to fetch file from ${f.url}`, {
          code: ErrorCodes.FileFetchFailed,
          cause: err instanceof Error ? err : void 0
        });
      }
      if (!res.ok) {
        throw new FluxerError(
          `Failed to fetch file from ${f.url}: ${res.status} ${res.statusText}`,
          {
            code: ErrorCodes.FileFetchFailed
          }
        );
      }
      const data = await res.arrayBuffer();
      result.push({ name: f.name, data, filename });
    } else if ("data" in f && f.data != null) {
      result.push({ name: f.name, data: f.data, filename });
    } else {
      throw new FluxerError(`File at index ${i} must have either "data" or "url"`, {
        code: ErrorCodes.InvalidAttachment
      });
    }
  }
  return result;
}
function toAPIAllowedMentions(options) {
  const result = {};
  if (options.parse?.length) result.parse = options.parse;
  if (options.users?.length) result.users = options.users;
  if (options.roles?.length) result.roles = options.roles;
  if (options.repliedUser !== void 0) result.replied_user = options.repliedUser;
  return result;
}
function applyReplyPingSuppression(body) {
  body.flags = (body.flags ?? 0) | MessageFlags.SuppressNotifications;
  body.allowed_mentions = {
    ...body.allowed_mentions ?? {},
    replied_user: false
  };
}
function buildSendBody(options) {
  const body = typeof options === "string" ? { content: options } : options;
  const result = {};
  if (body.content !== void 0) result.content = body.content;
  if (body.embeds?.length) {
    result.embeds = body.embeds.map((e) => e instanceof EmbedBuilder ? e.toJSON() : e);
  }
  if (body.files?.length && body.attachments) {
    result.attachments = body.attachments.map((a) => ({
      id: a.id,
      filename: a.filename,
      ...a.title != null && { title: a.title },
      ...a.description != null && { description: a.description },
      ...a.flags != null && { flags: a.flags }
    }));
  } else if (body.files?.length) {
    result.attachments = body.files.map((f, i) => ({
      id: i,
      filename: f.filename ?? f.name
    }));
  }
  if (body.allowedMentions) result.allowed_mentions = toAPIAllowedMentions(body.allowedMentions);
  if (body.stickerIds?.length) result.sticker_ids = body.stickerIds;
  if (body.nonce !== void 0) result.nonce = body.nonce;
  if (body.favoriteMemeId !== void 0) result.favorite_meme_id = body.favoriteMemeId;
  if (body.tts !== void 0) result.tts = body.tts;
  if (body.flags !== void 0) result.flags = body.flags;
  return result;
}
function toMessageReference(target) {
  const ref = {
    channel_id: target.channelId,
    message_id: target.messageId
  };
  if (target.guildId != null && target.guildId !== "") ref.guild_id = target.guildId;
  return ref;
}
async function prepareMessagePostPayload(options) {
  if (typeof options === "string") {
    if (options.length === 0) throw new RangeError("Cannot send an empty message");
    options = { content: options };
  }
  const { replyTo, ping, files, allowedMentions, ...sendFields } = options;
  const body = buildSendBody({ ...sendFields, files, allowedMentions });
  if (replyTo) {
    body.message_reference = toMessageReference(replyTo);
    const suppressPing = ping === false || allowedMentions?.repliedUser === false || body.allowed_mentions?.replied_user === false;
    if (suppressPing) applyReplyPingSuppression(body);
  } else if (allowedMentions?.repliedUser === false) {
    applyReplyPingSuppression(body);
  }
  const resolvedFiles = files?.length ? await resolveMessageFiles(files) : void 0;
  return { body, files: resolvedFiles };
}

// src/util/ReactionCollector.ts
import { EventEmitter } from "events";
import { Collection } from "@fluxerjs/collection";

// src/util/Events.ts
var Events = {
  Ready: "ready",
  MessageCreate: "messageCreate",
  MessageUpdate: "messageUpdate",
  MessageDelete: "messageDelete",
  MessageDeleteBulk: "messageDeleteBulk",
  MessageReactionAdd: "messageReactionAdd",
  MessageReactionRemove: "messageReactionRemove",
  MessageReactionRemoveAll: "messageReactionRemoveAll",
  MessageReactionRemoveEmoji: "messageReactionRemoveEmoji",
  InteractionCreate: "interactionCreate",
  GuildCreate: "guildCreate",
  GuildUpdate: "guildUpdate",
  GuildDelete: "guildDelete",
  GuildBanAdd: "guildBanAdd",
  GuildBanRemove: "guildBanRemove",
  GuildEmojisUpdate: "guildEmojisUpdate",
  GuildStickersUpdate: "guildStickersUpdate",
  GuildIntegrationsUpdate: "guildIntegrationsUpdate",
  GuildMemberAdd: "guildMemberAdd",
  GuildMemberUpdate: "guildMemberUpdate",
  GuildMemberRemove: "guildMemberRemove",
  GuildMembersChunk: "guildMembersChunk",
  GuildRoleCreate: "guildRoleCreate",
  GuildRoleUpdate: "guildRoleUpdate",
  GuildRoleDelete: "guildRoleDelete",
  GuildScheduledEventCreate: "guildScheduledEventCreate",
  GuildScheduledEventUpdate: "guildScheduledEventUpdate",
  GuildScheduledEventDelete: "guildScheduledEventDelete",
  ChannelCreate: "channelCreate",
  ChannelUpdate: "channelUpdate",
  ChannelDelete: "channelDelete",
  ChannelPinsUpdate: "channelPinsUpdate",
  InviteCreate: "inviteCreate",
  InviteDelete: "inviteDelete",
  TypingStart: "typingStart",
  UserUpdate: "userUpdate",
  PresenceUpdate: "presenceUpdate",
  VoiceStateUpdate: "voiceStateUpdate",
  VoiceServerUpdate: "voiceServerUpdate",
  VoiceStatesSync: "voiceStatesSync",
  WebhooksUpdate: "webhooksUpdate",
  Resumed: "resumed",
  Error: "error",
  Debug: "debug"
};

// src/util/ReactionCollector.ts
var ReactionCollector = class extends EventEmitter {
  client;
  messageId;
  channelId;
  options;
  collected = new Collection();
  _timeout = null;
  _ended = false;
  _listener;
  constructor(client, messageId, channelId, options = {}) {
    super();
    this.client = client;
    this.messageId = messageId;
    this.channelId = channelId;
    this.options = {
      filter: options.filter ?? (() => true),
      time: options.time ?? 0,
      max: options.max ?? 0
    };
    this._listener = (reaction, user, _msgId, chId, _emoji, userId) => {
      if (this._ended || reaction.messageId !== this.messageId || chId !== this.channelId) return;
      if (!this.options.filter(reaction, user)) return;
      const key = `${userId}:${reaction.emoji.id ?? reaction.emoji.name}`;
      this.collected.set(key, { reaction, user });
      this.emit("collect", reaction, user);
      if (this.options.max > 0 && this.collected.size >= this.options.max) {
        this.stop("limit");
      }
    };
    this.client.on(Events.MessageReactionAdd, this._listener);
    if (this.options.time > 0) {
      this._timeout = setTimeout(() => this.stop("time"), this.options.time);
    }
  }
  stop(reason = "user") {
    if (this._ended) return;
    this._ended = true;
    this.client.off(Events.MessageReactionAdd, this._listener);
    if (this._timeout) {
      clearTimeout(this._timeout);
      this._timeout = null;
    }
    this.emit("end", this.collected, reason);
  }
  on(event, listener) {
    return super.on(event, listener);
  }
  emit(event, ...args) {
    return super.emit(event, ...args);
  }
};

// src/structures/Message.ts
var Message = class _Message extends Base {
  client;
  id;
  channelId;
  guildId;
  author;
  content;
  createdAt;
  editedAt;
  pinned;
  attachments;
  type;
  flags;
  mentionEveryone;
  tts;
  embeds;
  stickers;
  reactions;
  messageReference;
  messageSnapshots;
  call;
  referencedMessage;
  /** Webhook ID if this message was sent via webhook. Null otherwise. */
  webhookId;
  /** Users mentioned in this message. */
  mentions;
  /** Role IDs mentioned in this message. */
  mentionRoles;
  /** Client-side nonce for acknowledgment. Null if not provided. */
  nonce;
  /**
   * Channel where this message was sent. Resolved from cache; null if not cached.
   * Messages can only exist in text-based channels (text, DM, announcement), so this always has send() when non-null.
   */
  get channel() {
    return this.client.channels.get(this.channelId) ?? null;
  }
  /** Guild where this message was sent. Resolved from cache; null for DMs or if not cached. */
  get guild() {
    return this.guildId ? this.client.guilds.get(this.guildId) ?? null : null;
  }
  /**
   * Resolve the channel (from cache or API). Use when you need the channel and it may not be cached.
   * @returns The channel
   * @throws FluxerError with CHANNEL_NOT_FOUND if the channel does not exist
   */
  async resolveChannel() {
    return this.client.channels.resolve(this.channelId);
  }
  /**
   * Resolve the guild (from cache or API). Returns null for DMs.
   * @returns The guild, or null if this is a DM or guild not found
   */
  async resolveGuild() {
    return this.guildId ? this.client.guilds.resolve(this.guildId) : null;
  }
  /** @param data - API message from POST/PATCH /channels/{id}/messages or gateway MESSAGE_CREATE */
  constructor(client, data) {
    super();
    this.client = client;
    this.id = data.id;
    this.channelId = data.channel_id;
    this.guildId = data.guild_id ?? null;
    this.author = client.getOrCreateUser(data.author);
    this.content = data.content;
    this.createdAt = new Date(data.timestamp);
    this.editedAt = data.edited_timestamp ? new Date(data.edited_timestamp) : null;
    this.pinned = data.pinned;
    this.attachments = new Collection2();
    for (const a of data.attachments ?? []) this.attachments.set(a.id, a);
    this.type = data.type ?? MessageType.Default;
    this.flags = data.flags ?? 0;
    this.mentionEveryone = data.mention_everyone ?? false;
    this.tts = data.tts ?? false;
    this.embeds = data.embeds ?? [];
    this.stickers = data.stickers ?? [];
    this.reactions = data.reactions ?? [];
    this.messageReference = data.message_reference ?? null;
    this.messageSnapshots = data.message_snapshots ?? [];
    this.call = data.call ?? null;
    this.referencedMessage = data.referenced_message ? new _Message(client, data.referenced_message) : null;
    this.webhookId = data.webhook_id ?? null;
    this.mentions = (data.mentions ?? []).map((u) => client.getOrCreateUser(u));
    this.mentionRoles = data.mention_roles ?? [];
    this.nonce = data.nonce ?? null;
  }
  /**
   * Send a message to this channel without replying. Use when you want a standalone message.
   * @param options - Text content or object with content, embeds, and/or files
   * @example
   * await message.send('Pong!');
   * await message.send({ embeds: [embed] }); // EmbedBuilder auto-converted
   * await message.send({ content: 'File', files: [{ name: 'data.txt', data }] });
   */
  async send(options) {
    const payload = await prepareMessagePostPayload(options);
    return this._send(payload);
  }
  /**
   * Send a message to a specific channel. Use for logging, forwarding, or sending to another channel in the guild.
   * @param channelId - Snowflake of the target channel (e.g. log channel ID)
   * @param options - Text content or object with content and/or embeds
   * @example
   * await message.sendTo(logChannelId, 'User ' + message.author.username + ' said: ' + message.content);
   * await message.sendTo(announceChannelId, { embeds: [embed] });
   */
  async sendTo(channelId, options) {
    return this.client.channels.send(channelId, options);
  }
  /**
   * Reply to this message (shows as a reply in the client).
   * @param options - Text content or object with content, embeds, and/or reply options (ping, replyTo)
   * @example
   * await message.reply('Pong!');
   * await message.reply({ embeds: [embed] });
   * await message.reply('No ping!', { ping: false });
   * await message.reply({ content: 'Silent reply', allowedMentions: AllowedMentions.suppressReplyPing });
   * await message.reply({ content: 'Reply to other', replyTo: otherMessage });
   */
  async reply(options, replyOptions) {
    const opts = typeof options === "string" ? { content: options } : options;
    const mergedReply = replyOptions ?? (opts.ping !== void 0 || opts.replyTo !== void 0) ? { ping: opts.ping, replyTo: opts.replyTo } : void 0;
    const refMessage = mergedReply?.replyTo ?? this;
    const replyTo = refMessage instanceof _Message ? {
      channelId: refMessage.channelId,
      messageId: refMessage.id,
      guildId: refMessage.guildId
    } : refMessage;
    const { replyTo: _ignored, ping: _ping, ...sendOpts } = opts;
    const payload = await prepareMessagePostPayload({
      ...sendOpts,
      replyTo,
      ping: mergedReply?.ping
    });
    return this._send(payload);
  }
  /** Exposed for testing — prefer `message.reply()` / `prepareMessagePostPayload()`. */
  static async _createMessageBody(content, referenced_message, ping) {
    if (typeof content === "string" && content.length === 0) {
      throw new RangeError("Cannot send an empty message");
    }
    if (typeof content === "string") {
      const opts2 = { content };
      if (referenced_message) {
        opts2.replyTo = {
          channelId: referenced_message.channel_id,
          messageId: referenced_message.message_id,
          ...referenced_message.guild_id != null ? { guildId: referenced_message.guild_id } : {}
        };
        if (ping !== void 0) opts2.ping = ping;
      }
      return prepareMessagePostPayload(opts2);
    }
    const opts = { ...content };
    if (referenced_message) {
      opts.replyTo = {
        channelId: referenced_message.channel_id,
        messageId: referenced_message.message_id,
        ...referenced_message.guild_id != null ? { guildId: referenced_message.guild_id } : {}
      };
      if (ping !== void 0) opts.ping = ping;
    }
    return prepareMessagePostPayload(opts);
  }
  async _send(payload) {
    const data = await this.client.rest.post(
      Routes.channelMessages(this.channelId),
      payload
    );
    this.client._addMessageToCache(this.channelId, data);
    return new _Message(this.client, data);
  }
  /**
   * Edit this message. Only the author (or admins) can edit.
   * @param options - New content and/or embeds
   */
  async edit(options) {
    const body = {};
    if (options.content !== void 0) body.content = options.content;
    if (options.embeds?.length) {
      body.embeds = options.embeds.map((e) => e instanceof EmbedBuilder2 ? e.toJSON() : e);
    }
    if (options.allowedMentions) {
      body.allowed_mentions = toAPIAllowedMentions(options.allowedMentions);
    }
    const data = await this.client.rest.patch(Routes.channelMessage(this.channelId, this.id), {
      body
    });
    return new _Message(this.client, data);
  }
  /**
   * Create a reaction collector for this message.
   * Collects reactions matching the filter until time expires or max is reached.
   * @param options - Filter, time (ms), and max count
   * @example
   * const collector = message.createReactionCollector({ filter: (r, u) => u.id === userId, time: 10000 });
   * collector.on('collect', (reaction, user) => console.log(user.username, 'reacted with', reaction.emoji.name));
   * collector.on('end', (collected, reason) => { ... });
   */
  createReactionCollector(options) {
    return new ReactionCollector(this.client, this.id, this.channelId, options);
  }
  /**
   * Re-fetch this message from the API to get the latest content, embeds, reactions, etc.
   * Use when you have a stale Message (e.g. from an old event or cache) and need fresh data.
   * @returns The updated message
   * @throws FluxerError with MESSAGE_NOT_FOUND if the message was deleted or does not exist
   * @example
   * const updated = await message.fetch();
   * console.log('Latest content:', updated.content);
   */
  async fetch() {
    return this.client.channels.fetchMessage(this.channelId, this.id);
  }
  /** Delete this message. */
  async delete() {
    await this.client.rest.delete(Routes.channelMessage(this.channelId, this.id));
  }
  /**
   * Delete a specific attachment from this message.
   * DELETE /channels/{id}/messages/{id}/attachments/{attachmentId}.
   */
  async deleteAttachment(attachmentId) {
    await this.client.rest.delete(
      Routes.channelMessageAttachment(this.channelId, this.id, attachmentId),
      { auth: true }
    );
    this.attachments.delete(attachmentId);
  }
  /** Pin this message to the channel. Requires Manage Messages permission. */
  async pin() {
    await this.client.rest.put(Routes.channelPinMessage(this.channelId, this.id));
    this.pinned = true;
  }
  /** Unpin this message from the channel. Requires Manage Messages permission. */
  async unpin() {
    await this.client.rest.delete(Routes.channelPinMessage(this.channelId, this.id));
    this.pinned = false;
  }
  /**
   * Format emoji for reaction API: unicode string or "name:id" for custom.
   * For string resolution (e.g. :name:), use client.resolveEmoji; Message methods resolve automatically when guildId is available.
   */
  static formatEmoji(emoji) {
    if (typeof emoji === "string") return emoji;
    return `${emoji.name}:${emoji.id}`;
  }
  resolveEmojiForReaction(emoji) {
    return this.client.resolveEmoji(emoji, this.guildId);
  }
  /**
   * Add a reaction to this message (as the bot).
   * @param emoji - Unicode emoji, custom `{ name, id }`, `:name:`, `name:id`, or `<:name:id>`
   */
  async react(emoji) {
    const emojiStr = await this.resolveEmojiForReaction(emoji);
    const route = `${Routes.channelMessageReaction(this.channelId, this.id, emojiStr)}/@me`;
    await this.client.rest.put(route);
  }
  /**
   * Remove the bot's reaction, or a specific user's reaction if userId is provided.
   * @param emoji - Unicode emoji, custom `{ name, id }`, `:name:`, `name:id`, or `<:name:id>`
   * @param userId - If provided, removes that user's reaction (requires moderator permissions)
   */
  async removeReaction(emoji, userId) {
    const emojiStr = await this.resolveEmojiForReaction(emoji);
    const route = `${Routes.channelMessageReaction(this.channelId, this.id, emojiStr)}/${userId ?? "@me"}`;
    await this.client.rest.delete(route);
  }
  /**
   * Remove all reactions from this message.
   * Requires moderator permissions.
   */
  async removeAllReactions() {
    await this.client.rest.delete(Routes.channelMessageReactions(this.channelId, this.id));
  }
  /**
   * Remove all reactions of a specific emoji from this message.
   * @param emoji - Unicode emoji, custom `{ name, id }`, `:name:`, `name:id`, or `<:name:id>`. Requires moderator permissions.
   */
  async removeReactionEmoji(emoji) {
    const emojiStr = await this.resolveEmojiForReaction(emoji);
    await this.client.rest.delete(Routes.channelMessageReaction(this.channelId, this.id, emojiStr));
  }
  /**
   * Fetch users who reacted with the given emoji.
   * @param emoji - Unicode emoji or custom `{ name, id }`
   * @param options - limit (1–100), after (user ID for pagination)
   * @returns Array of User objects
   */
  async fetchReactionUsers(emoji, options) {
    const emojiStr = await this.resolveEmojiForReaction(emoji);
    const params = new URLSearchParams();
    if (options?.limit != null) params.set("limit", String(options.limit));
    if (options?.after) params.set("after", options.after);
    const qs = params.toString();
    const route = Routes.channelMessageReaction(this.channelId, this.id, emojiStr) + (qs ? `?${qs}` : "");
    const data = await this.client.rest.get(route);
    const list = Array.isArray(data) ? data : data?.users ?? [];
    return list.map((u) => this.client.getOrCreateUser(u));
  }
};
export {
  Message
};
